[llvm-exegesis] Use explicit error classes for different snippet crashes

boomanaiden154 opened the pull request:
This patch switches to using explicit snippet crashes that contain more information about the specific type of error (like the address for a segmentation fault) that occurred. All these new error classes inherit from SnippetExecutionFailure to allow for easily grabbing all of them in addition to filtering for specific types using the standard LLVM error primitives.

boomanaiden154 commented:
Bump on this when reviewers have a chance. Thanks!

legrosbuffle reviewed (commented):
Sorry forgot to hit send.

legrosbuffle commented on llvm/tools/llvm-exegesis/lib/Error.h:
```diff
@@ -41,14 +43,19 @@ class ClusteringError : public ErrorInfo<ClusteringError> {
 public:
   static char ID;
-  SnippetCrash(const Twine &S) : Msg(S.str()) {}
+  SnippetCrash(const Twine &S)
```
All right, sorry for the conflicting recommendations. It's hard to know exactly what the best approach is given that I can't see how you want to use it downstream. 

How are you accessing the signal and address if there are no getters for these ?

legrosbuffle replied:
Coming back to that patch to try to see whay I made that remark: I think that this was to avoid the `if (!E.isA<SnippetCrash>() && !E.isA<SnippetSegfaultCrash>())`. This is kind of related to https://github.com/llvm/llvm-project/pull/74211.

boomanaiden154 replied:
Ah, makes sense.

Either way I think we would need a conditional there to send appropriate diagnostics to the user, but a generic separate error would be more scalable than individual error classes. I'm not sure there's a large benefit to separating out the error classes though, especially given it would end up looking pretty similar to D158052. What do you think is best in this situation?

legrosbuffle replied:
> https://github.com/llvm/llvm-project/pull/74211 adds additional support (including a getter). https://github.com/google/gematria/pull/30 shows the usage on top of that PR.

Thanks for the pointers. With the added getter in #74211 I think we should probably have two different types, because the getter would be meaningless in case of a failed child process without a crash. In particular, l137 in https://github.com/google/gematria/pull/30 would be incorrect. WDYT ?

boomanaiden154 replied:
That's true. It makes the assumption that every crash is a segfault which is not correct. I'll refactor it to a separate error.

boomanaiden154 committed:
[llvm-exegesis] Use explicit error classes for different snippet crashes

This patch switches to using explicit snippet crashes that contain more
information about the specific type of error (like the address for a
segmentation fault) that occurred. All these new error classes inherit
from SnippetExecutionFailure to allow for easily grabbing all of them in
addition to filtering for specific types using the standard LLVM error
primitives.

legrosbuffle commented on llvm/tools/llvm-exegesis/lib/BenchmarkRunner.cpp:
```diff
@@ -141,17 +141,10 @@ class InProcessFunctionExecutorImpl : public BenchmarkRunner::FunctionExecutor {
-        }
-#endif
-        return make_error<SnippetCrash>(std::move(Msg));
+        return make_error<SnippetSignal>(CRC.RetCode - kSigOffset);
```
Is the signal meaningful on windows ?

boomanaiden154 replied:
No. I've readded the preprocessor conditional here for Unix vs non-unix systems.

legrosbuffle commented on llvm/tools/llvm-exegesis/lib/BenchmarkRunner.cpp:
```diff
@@ -575,7 +570,7 @@ Expected<Benchmark> BenchmarkRunner::runConfiguration(
 
   if (Error E = NewMeasurements.takeError()) {
-    if (!E.isA<SnippetCrash>())
+    if (!E.isA<SnippetExecutionFailure>())
```
You should probably rebase.

legrosbuffle reviewed (commented):
Design looks good to me; just a couple remarks on the code.

boomanaiden154 committed:
Only return signals on Unix
